Collect Fluentbit logs

RaulGracia opened the pull request:
**Change log description**
Added a new log collecting script for fluentbit service (using syslog output plugin).

**Purpose of the change**
Fixes #8 and #9.

**What the code does**
This PR adds a new script for collecting logs from a Kubernetes cluster containing a `fluentbit` service. This script is useful in scenarios where `fluentbit` stores logs from cluster pods locally (either in local storage or in a mounted volume) in the form of log files. The script does the following work:
1) It loops though the existing `fluentbit` pods and creates a temporary `tar.gz` file on each pod with all the logs.
2) It download these compressed log files from all the `fluentbit` pods and tries to structure the folders in a meaningful way for developers inspecting the logs.
3) As in the existing script `pravega-k8s-log.sh`, the new one also `describe`s all the pods, pvcs, and services in the cluster.
4) It cleans up temporary files both locally and in `fluentbit` pods.

The `README` has been updated as well.

**How to verify it**
Collected logs from 2 clusters successfully. One of these clusters contained 4 `fluentbit` pods.

RaulGracia committed:
Issue 8: First version of script to collect fluentbit logs.

Signed-off-by: Raúl Gracia <raul.gracia@emc.com>

RaulGracia committed:
Issue 8: Fixed temporal log storage for multiple fluentbit instances.

Signed-off-by: Raúl Gracia <raul.gracia@emc.com>

RaulGracia committed:
Issue 8: Renamed variable for clarity.

Signed-off-by: Raúl Gracia <raul.gracia@emc.com>

RaulGracia committed:
Issue 8: Updated README.

Signed-off-by: Raúl Gracia <raul.gracia@emc.com>

RaulGracia commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Improve log collection in k8 environment with Fluentd](https://trello.com/c/TOLzmTxK/110-improve-log-collection-in-k8-environment-with-fluentd)

RaulGracia commented:
@sumit-bm @deenav @anishakj please, provide your feedback on the PR.

sumit-bm commented:
sorry, closed by mistake.

sumit-bm commented:
Hi @RaulGracia

@deenav and I went over the implementation and we have few comments on this, we are still collating more comprehensive review areas but i think we can take this as first review and will update more on this thread.

1. Looks like the fluent-bit log collection doesn't take care restarted pods logs, (while restarted pods have logs under `/var/log/containers` as n.log [n represent restart count] whereas `/var/vcap/store/docker/docker/containers/` contains rotated x.json.log

2. Do we need all the files from `/var/vcap/store/docker/docker/containers/` or just x.json.log is fine? Currently utility is taking all files like checkpoint / hostconifg.json etc from the above position, which doesn't have much useful data

3. Looks like Fluent log is not taking complete rotated logs as well as compared to all rotated logs present inside the pod. Like under controller pod all rotated logs are present, but fluent is having some rotated logs but not all old logs. 

4. Could logs from Evicted pods be available using Fluent?

5. If a pod is failed and removed (non replica set), Fluent log doesn't contain a log for the same.

Thanks,
-Deena, Sumit.

RaulGracia committed:
Issue 8: Considered the case in which the logs directory for a pod does not exist (e.g., pod is in CrashLoopBackOff).

Signed-off-by: Raúl Gracia <raul.gracia@emc.com>

RaulGracia committed:
Issue 8: Cleanup any possible existing temporal compressed file in fluentbit pods. Delete from the result tar package all non .log files.

Signed-off-by: Raúl Gracia <raul.gracia@emc.com>

RaulGracia commented:
Hi @sumit-bm,
I addressed 3 points in the script:
1) Handle the case in which a log directory for a pod does not exist (e.g., pod is in CrashLoopBackOff).
2) Initial cleanup of any existing temporal tar file in fluentbit pods.
3) Cleanup of all non `.log` files from the resulting packaged tar.

I feel that some of your comments are more related to the way fluentbit is logging rather than with problems of the script, right? I don't know if we can do much from a collection script perspective. Maybe we will need to check the actual configuration of the fluentbit service to answer some of these questions ([see this repo](https://github.com/pivotal-cf/fluent-bit-out-syslog)).

Please, for the next round, it would be great that reviewers could provide a regular review accepting or not the PR changes, and specifying the issues found in the script file. Thanks!

deenav commented on pravega-k8-log-collection/pravega-k8s-fluentbit-log.sh:
```diff
@@ -0,0 +1,93 @@
+        done
+        # Clean temporal files.
+        rm -rf $output_dir/$fluentbit_pod.tar.gz
+        rm -rf $output_dir/var
```
It's nice to cleanup the temporary files like pod-lognames & container-lognames
```suggestion
        rm -rf $output_dir/var
        rm -rf $output_dir/pod-lognames
        rm -rf $output_dir/container-lognames
```

deenav reviewed (changes requested):
For restarted pods, `ls /var/log/containers/` will have more than one soft link pointing to different directories (container ids) in `/var/vcap/store/docker/docker/containers/`. 
From the below code block `log_pod_name` will be same for the restarted pods but where as the respective `log_container_name` would be different for each restart. 
So, when you are moving `$output_dir/$fluentbit_pod/$log_container_name` to   `$output_dir/$fluentbit_pod/$log_pod_name`, previous `log_container_name` gets overwritten with the latest for the restarted pods.

```
# Relate the name of the pods being logged with the actual log data.
        kubectl -n $namespace exec $fluentbit_pod -- ls /var/vcap/store/docker/docker/containers/ > $output_dir/container-lognames
        kubectl -n $namespace exec $fluentbit_pod -- ls /var/log/containers/ > $output_dir/pod-lognames
        for podlog in $(cat $output_dir/pod-lognames)
        do
                log_pod_name=$(echo $podlog | awk -F_ '{print $1}')
                log_container_name=$(echo $podlog | awk -F- '{print $NF}' |  awk -F. '{print $1}')
                # This will help to easily locate the logs by pod name.
                mv $output_dir/$fluentbit_pod/$log_container_name  $output_dir/$fluentbit_pod/$log_pod_name || true
```

It would be good to have a better method for relating the name of the pods with container id's which will fix the issue of restarted pods logs are missing in the current version of this script.

shrids commented on pravega-k8-log-collection/pravega-k8s-fluentbit-log.sh:
```diff
@@ -0,0 +1,95 @@
+        set +e
+        kubectl -n $namespace exec $fluentbit_pod -- bash -c "tar -czf /tmp/$fluentbit_pod.tar.gz /var/vcap/store/docker/docker/containers"
+        exitcode=$?
+        # Some logs may be being written during collection, which makes tar to exit with a non-zero code. Need to handle this case.
```
We could use `--ignore-failed-read ` option of tar to solve the error that logs may be being written during collection.

RaulGracia replied:
Tried with that flag, but `tar` seems to continue exiting with with non-zero code. That is, using:
`kubectl -n $namespace exec $fluentbit_pod -- bash -c "tar --ignore-failed-read -czf /tmp/$fluentbit_pod.tar.gz /var/vcap/store/docker/docker/containers"`
If we remove the non-zero exit code handling, we get the following error:
```
tar: /var/vcap/store/docker/docker/containers/43b58b96ef7f55459261247c397acfa72e26b3e7c055909ea199ed7df4aa3d43/43b58b96ef7f55459261247c397acfa72e26b3e7c055909ea199ed7df4aa3d43-json.log: file changed as we read it
command terminated with exit code 1
```
I assume that the `--ignore-failed-read` is supposed to help us to remove the non-zero exit code handling, or maybe I'm totally wrong here.

shrids reviewed (approved):
+1 LGTM apart from one minor comment.

RaulGracia committed:
Issue 8: Added container id prefix to avoid overwriting logs from restarted pods.

Signed-off-by: Raúl Gracia <raul.gracia@emc.com>

RaulGracia commented:
Good catch, @deenav! Yes, you are right, we were overwriting logs from restarted pods. To avoid that, now I add to the pod name the 5 first digits of the container name. The result is the following:
1) We can see a restarted segment store pod:
```
kubectl get pods
NAME                                                    READY   STATUS    RESTARTS   AGE
benchmark-pod                                           1/1     Running   0          16d
kilted-magpie-nfs-client-provisioner-58c457bb6d-zbp4c   1/1     Running   1          16d
pravega-bookie-0                                        1/1     Running   0          22h
pravega-bookie-1                                        1/1     Running   0          22h
pravega-bookie-2                                        1/1     Running   0          22h
pravega-operator-648f658d8f-vdqgp                       1/1     Running   0          22h
pravega-pravega-controller-846fd98cff-bzkc9             1/1     Running   0          22h
pravega-pravega-segmentstore-0                          1/1     Running   0          22h
pravega-pravega-segmentstore-1                          1/1     Running   0          22h
pravega-pravega-segmentstore-2                          1/1     Running   1          22h  <- this one
```
2) When executing the script, now we have:
```
ls fluent-bit-ffw5m/
32fd7e7109b12f7f733be4f65edb1d21eb8d5384a71c008531276fc526fcbc67  c8c780d43496f8cefc6f98a3e5a880261c7fab807c9b10a8466696bac0942d87  kubernetes-dashboard-5f4b59b97f-fqmdh-43b58  zookeeper-operator-5b4f86b64b-zdgsv-0f1c4
4bacdc158eed77bea5076e445d7246d16741b815a69010c27a5775b815e5b142  fluent-bit-ffw5m-3ca0e                                            monitoring-influxdb-cdcf4674-clx7p-5b8fb
85847412331bb82f897bd59f52e330d8d3dde1b66fac65fc03428039bb2ed50f  fluent-bit-ffw5m-76259                                            pravega-pravega-segmentstore-2-49e69
b1c56d8d02ff3640a65f245039697c6bdedca35865a8f806cb9673208ba5b3f8  fluent-bit-ffw5m-b1524                                            pravega-pravega-segmentstore-2-af73a
```
As you can see, now there are 2 folders containing the logs for `segmentstore-2`. Added the container prefix to the pod name solves the overwriting problem.

deenav reviewed (approved):
Verified that restarted Pod logs are available now.

adrianmo commented on pravega-k8-log-collection/pravega-k8s-fluentbit-log.sh:
```diff
@@ -0,0 +1,94 @@
+# We now rok only on the "fluentbit" pods that contain the logs of the rest of services.
+for i in $(cat $output_dir/kubectl-get-po-svc-pvc-all-namespace_output | grep -i fluent-bit)
+do
+
```
Everything inside this `for` block needs to be indented.

RaulGracia replied:
Done.

adrianmo commented on pravega-k8-log-collection/pravega-k8s-fluentbit-log.sh:
```diff
@@ -0,0 +1,94 @@
+
+    # Collect logs from "Running" pods, if pods are not in "Running" state it will collect describe information.
+    if [ "`echo $i | awk '{print $4}'`" == "Running" ]; then
+    {
```
Are these braces `{` valid in shell scripts? I haven't seen them before.

RaulGracia replied:
Works, but I removed them.

adrianmo commented on pravega-k8-log-collection/pravega-k8s-fluentbit-log.sh:
```diff
@@ -0,0 +1,94 @@
+        # Clean any pre-existing tar file in the fluentbit pod.
+        kubectl -n $namespace exec $fluentbit_pod -- bash -c "rm /tmp/$fluentbit_pod.tar.gz" || true
+        # Create a tar.gz file with all the logs contained in the fluentbit pod.
+        set +e
```
To avoid having to unset `-e` temporarily, can you do the following:

```
exitcode=0
kubectl -n $namespace exec $fluentbit_pod -- bash -c "tar -czf /tmp/$fluentbit_pod.tar.gz /var/vcap/store/docker/docker/containers" || exitcode=$?
```

RaulGracia replied:
Done.

adrianmo commented on pravega-k8-log-collection/pravega-k8s-fluentbit-log.sh:
```diff
@@ -0,0 +1,94 @@
+        # Remove temporal compressed file from fluentbit pod.
+        kubectl -n $namespace exec $fluentbit_pod -- bash -c "rm /tmp/$fluentbit_pod.tar.gz"
+        # Untar the contents to provide a better structure of the resulting logs artifact.
+        tar -xzvf $output_dir/$fluentbit_pod.tar.gz -C $output_dir
```
Maybe we want to remove the `v` flag to reduce verbosity and avoid unnecessary logs.

RaulGracia replied:
Done.

adrianmo commented on pravega-k8-log-collection/pravega-k8s-fluentbit-log.sh:
```diff
@@ -0,0 +1,94 @@
+                log_container_name=$(echo $podlog | awk -F- '{print $NF}' |  awk -F. '{print $1}')
+                # This will help to easily locate the logs by pod name, but also appended the container prefix (5 chars)
+                # to avoid overwriting logs from restarted pods.
+                mv $output_dir/$fluentbit_pod/$log_container_name  $output_dir/$fluentbit_pod/$log_pod_name'-'$(echo $log_container_name | awk '{print substr($0,0,5)}') || true
```
Is this command allowed or expected to fail?

RaulGracia replied:
As @deenav noticed, it is expected to fail in the case that the pod is in `CrashLoopBackOff` state.

adrianmo commented on pravega-k8-log-collection/pravega-k8s-fluentbit-log.sh:
```diff
@@ -0,0 +1,94 @@
+        rm -rf $output_dir/var
+    }
+    else
+       echo $fluentbit_pod" is not in running state"
```
Is this variable ever set in this context?

RaulGracia replied:
Fixed.

adrianmo commented on pravega-k8-log-collection/pravega-k8s-fluentbit-log.sh:
```diff
@@ -0,0 +1,94 @@
+# Create compressed log archive with date stamp
+tar zcf $cwd/pravega-k8s-logs/pravega-k8s-logging-`date '+%Y-%m-%d-%H-%M-%S'`.tgz *
+rm -rf ../$output_dir
+echo "Final logs are available in `ls $cwd/pravega-k8s-logs/*.tgz`"
```
It'd be good to add an `exit 0` at the end to indicate that the script has run successfully.

RaulGracia replied:
Done.

adrianmo commented on pravega-k8-log-collection/pravega-k8s-fluentbit-log.sh:
```diff
@@ -0,0 +1,94 @@
+set -euo pipefail
+# Setting up temporary log collection directory under /tmp/logdir
+mkdir -p ./logdir
+output_dir="./logdir"
```
These commands can be inverted and make `mkdir` use the `output_dir` variable.

RaulGracia replied:
Done.

adrianmo reviewed (changes requested):
Thanks @RaulGracia, the script looks nice. I've added a few minor comments.

sumit-bm reviewed (approved):
+1

RaulGracia commented:
I have addressed all your comments and tested the script, and it works. Now we can use it for a couple of days in Jenkins, and if everything works as expected, then we could merge this PR. Any improvement on the script can be done in a separate PR.

RaulGracia committed:
Issue 8: Improvements from PR comments.

Signed-off-by: Raúl Gracia <raul.gracia@emc.com>

RaulGracia committed:
Issue 8: Improvements from PR comments.

Signed-off-by: Raúl Gracia <raul.gracia@emc.com>

adrianmo reviewed (approved):
Thanks @RaulGracia. The script looks good!

anishakj commented:
Script is working fine with Jenkins job.
